Create a service that hosts ephemeral git repositories

faec opened the pull request:
This is a server that hosts ephemeral git repositories, intended for applying code intelligence to web pages and similar short-term sources.

Helps https://github.com/sourcegraph/sourcegraph/issues/423

Server parameters and testing procedure are documented in `README.md`.

faec commented:
Converging on a real version; main thing left is better error handling / input checks, and adding tests.

nicksnyder commented:
lmk when this is ready for review. I agree that in-memory sounds like overkill at this point.

nicksnyder commented on cmd/web-repo-proxy/main.go:
```diff
@@ -0,0 +1,323 @@
+}
+
+var (
+	repositories            = map[string]*repository{}
```
Can you explain the benefits of having this in-memory map?

It feels like there is a lot of complexity, locking, and edge cases around keeping this synced with what is on disk.

Would it be possible to delete this entirely and just use disk as a source of truth for what exists, when things were created (and implicitly when they should be evicted).

faec replied:
It could be done, though with some inconvenience -- the repository name needs to be read in order to parse the URL path, so that would need to be represented on disk and fetched whenever we get a file request. What were you thinking for eviction? Just periodically poll all repositories under the root directory checking for an old creation date?

nicksnyder replied:
At a high level, creating a repo is as simple as dumping it to disk at a location like
```
/repositories/00001-foobar/file1
/repositories/00001-foobar/file2
/repositories/00002-blue/file3
/repositories/00003-/file4 // no name given
```

Then when you need to serve a file, you know exactly where to read it from (and if it is no longer on disk because it has been evicted then you will obviously get an error).

Eviction would just poll the /repositories directory looking for subdirs with old creation timestamps and delete them.

nicksnyder commented:
Can you update the PR description to:
1. reflect the final state (i.e. no pending TODOs)
2. Document a simple test plan (e.g. run on localhost, create a repo with this command `curl ....`, clone a repo like this `git clone ....`

faec commented:
Rewrote to use entirely on-disk state and remove all locking / caches, and addressed the remaining comments above.

Should this PR update the changelog? It doesn't seem "user-facing" in its current form but lmk if I should add a line for it.

nicksnyder commented on cmd/web-repo-proxy/main.go:
```diff
@@ -0,0 +1,361 @@
+func (r *repository) urlRoot() string {
+	root := "/repository/" + r.key
+	if r.name != "" {
+		return root + "/" + r.name
```
I think the URL path should be consistent with the filePathRoot above instead of using a slash here. Another benefit is that a client who clones this repo (i.e. our gitserver) will clone it into a unique path by default instead of potentially clobbering.

nicksnyder commented on cmd/web-repo-proxy/main.go:
```diff
@@ -0,0 +1,361 @@
+			return
+		}
+		repoKey := pathComponents[1]
+		repositories := repositoriesForKey(repoKey)
```
I think we should be able to eliminate this precheck. The static file handler should just return a 404 if the repo or file doesn't exist:
https://golang.org/src/net/http/fs.go?#L628

nicksnyder commented on cmd/web-repo-proxy/main.go:
```diff
@@ -0,0 +1,361 @@
+			return nil, err
+		}
+		repositoryRoot = newRoot
+		// Corner case: only finish if we picked a unique key.
```
If you take my other suggestion to flatten the directory structure, then I think this logic becomes simpler (and you can delete repositoriesForKey).

The sketch would looks something like:
1. Create the repo in a tmp dir (e.g. `/tmp/asdf`)
2. Generate a random key (e.g. `1234`)
3. Atomically move the tmp dir into the permanent place (e.g. `mv /tmp/asdf /repositories/reponame-1234`)

If the move succeeds, then you are done. If it fails, generate a new random number and retry a max of 3 times (if it still fails return the last error).

nicksnyder commented on cmd/web-repo-proxy/main.go:
```diff
@@ -0,0 +1,361 @@
+			// This repository has expired.
+			tempDir, err := ioutil.TempDir("", "web-repo-proxy")
+			if err != nil {
+				log15.Error("Couldn't create temporary directory", "error", err)
```
You should still attempt to os.RemoveAll in all error cases.

faec replied:
Perhaps I misunderstood, do you mean I should os.RemoveAll even if TempDir itself returns error? I thought this was resolved by the deferred RemoveAll on error (which you've left other comments about below).

nicksnyder commented on cmd/web-repo-proxy/main.go:
```diff
@@ -0,0 +1,361 @@
+func sanitizedCreateRequest(requestBody io.Reader) (*CreateRepositoryRequest, error) {
+	decoder := json.NewDecoder(requestBody)
+	var request CreateRepositoryRequest
+	err := decoder.Decode(&request)
```
no need for the intermediate decoder variable

```suggestion
	err := json.NewDecoder(requestBody).Decode(&request)
```

nicksnyder commented on cmd/web-repo-proxy/main_test.go:
```diff
@@ -0,0 +1,208 @@
+	}
+	response := CreateRepositoryResponse{}
+	responseDecoder := json.NewDecoder(result.Body)
+	err := responseDecoder.Decode(&response)
```
no need for the intermediate responseDecoder variable

```suggestion
	err := json.NewDecoder(result.Body).Decode(&response)
```

nicksnyder commented on cmd/web-repo-proxy/main.go:
```diff
@@ -0,0 +1,361 @@
+		// Send successful response to client.
+		response := CreateRepositoryResponse{
+			UrlPath:   repo.urlRoot(),
+			GoodUntil: goodUntil.Unix()}
```
nit: put `}` on next line

nicksnyder commented on cmd/web-repo-proxy/README.md:
```diff
@@ -0,0 +1,59 @@
+
+# Testing
+
+To try it out locally:
```
Thanks for documenting this! It would be cool if there were a single integration test (in addition to any other tests that you want) that actually started a server, created a repo, cloned it, and then verified the state of the repo.

faec commented:
Addressed all comments on main file, finishing tests now.

slimsag commented:
> is only run when go test is given the "-endToEnd" flag. Let me know if you have a preferred way of gating tests like that.

We usually put things like this behind a `testing.Short()` gate, e.g.:

https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/db/access_tokens_test.go#L229-231

slimsag commented on cmd/web-repo-proxy/main.go:
```diff
@@ -0,0 +1,321 @@
+
+	listenerPort = env.Get("LISTENER_PORT", "4014", "The network port to listen on")
+
+	maxRequestSize = 0x10000 // Repo creation requests must be at most 64KB.
```
Q: Why hexadecimal here and not just an integer? The latter seems like it would be more readable in general, or am I wrong?

slimsag replied:
e.g. `1024 * 64`

ijsnow commented on cmd/web-repo-proxy/main.go:
```diff
@@ -0,0 +1,324 @@
+		// Send successful response to client.
+		response := CreateRepositoryResponse{
+			URLPath:   repo.urlPath(),
+			GoodUntil: goodUntil.Unix(),
```
And I'm aware that this may not be in scope of this project or that we may want to handle this in the browser extension. I'm just saying what would be nice from the browser extension's perspective :)

nicksnyder replied:
Thanks for the info. You are right that the browser extension doesn't really need goodUntil, but it is useful for debugging, so I think it is indeed better to return it as a string (as that is more natural for a human, yet still parseable if we need).

The path returned will be able to be mapped directly to a repo name as you want, so I don't think we need to change anything here for now.

@faec in short, just update this to be formatted as a RFC3339 string.

faec commented on cmd/web-repo-proxy/main_test.go:
```diff
@@ -0,0 +1,285 @@
+	defer cancel()
+
+	// Delay to let the server start listening.
+	time.Sleep(time.Second / 10)
```
It doesn't speed the test up much since the main bottleneck is building the server, but it does make me feel less nervous to be doing it the right way ^^ done

ggilmore commented on cmd/web-repo-proxy/main.go:
```diff
@@ -0,0 +1,323 @@
+
+// goodUntil checks the repository's directory on disk and returns the time
+// it will expire, or an error if the repository couldn't be read.
+func (r repository) goodUntil() (time.Time, error) {
```
`expiration()` seems like a more straightforward+descriptive name for this funciton

ggilmore commented on cmd/web-repo-proxy/main.go:
```diff
@@ -0,0 +1,323 @@
+	for _, command := range gitCommands {
+		cmd := exec.Command("git", command...)
+		cmd.Dir = tempRoot
+		if err = cmd.Run(); err != nil {
```
The `err` returned by `cmd.Run()` typically won't have a very helpful error message (something like `exit status 1`. You can use `cmd.CombinedOutput()` to capture the `stderr` output and include it in your printed error message. 

See https://stackoverflow.com/a/18177139 for more info.

ggilmore commented on cmd/web-repo-proxy/main.go:
```diff
@@ -0,0 +1,323 @@
+		return
+	}
+	for _, file := range files {
+		if file.ModTime().Add(repositoryTTL).Before(curTime) {
```
I think this line would be easier to read if it was written more like:

```go
expiration := file.ModTime().add(repositoryTTL)
if time.Now().After(expiration) {
   ...
}
```

ggilmore commented on cmd/web-repo-proxy/main.go:
```diff
@@ -0,0 +1,323 @@
+			// This repository has expired.
+			tempDir, err := ioutil.TempDir("", "web-repo-proxy")
+			if err != nil {
+				log15.Error("Couldn't create temporary directory", "error", err)
```
To keep this log consistent with the other log at the end of the function:

```suggestion
				log15.Error("Couldn't create temporary directory", "tempDir", tempDir, "error", err)
```

ggilmore commented on cmd/web-repo-proxy/main.go:
```diff
@@ -0,0 +1,323 @@
+			}
+			err = os.RemoveAll(tempDir)
+			if err != nil {
+				log15.Error("Couldn't delete temporary directory", "tempDir", tempDir)
```
To keep this log consistent with the other log above for creating the temporary directory: 

```suggestion
				log15.Error("Couldn't delete temporary directory", "tempDir", tempDir, "error", err)
```

ggilmore commented on cmd/web-repo-proxy/main.go:
```diff
@@ -0,0 +1,323 @@
+
+// Check whether any repositories have expired, and delete them if so.
+func deleteExpiredRepositories() {
+	curTime := time.Now()
```
For `curTime`, I suggest:

1) Move `curTime`'s declaration to right above the `for` loop so that it can be closer to where it's actually used. See: https://github.com/sourcegraph/sourcegraph/blob/master/doc/dev/style.md#group-code-blocks-logically

2) Change `curTime`'s name to just `now`. e.g. `now := time.Now()`. See https://github.com/sourcegraph/sourcegraph/blob/master/doc/dev/style.md#use-pronouncable-names

ggilmore commented on cmd/web-repo-proxy/main.go:
```diff
@@ -0,0 +1,323 @@
+				continue
+			}
+			defer os.RemoveAll(tempDir)
+			fromPath := filepath.Join(repositoriesRoot, file.Name())
```
I would prefer removing the `Path` suffix from these variable names. Example:

```go 
from := filepath.Join(repositoriesRoot, file.Name())
to := filepath.Join(tempDir, file.Name())
err := os.Rename(from, to)
```

_`old` and `new` might also work._

See https://github.com/sourcegraph/sourcegraph/blob/master/doc/dev/style.md#keep-names-short

ggilmore commented on cmd/web-repo-proxy/main.go:
```diff
@@ -0,0 +1,323 @@
+// their goodUntil field is past.
+func repositoryDeleter() {
+	for {
+		deleteExpiredRepositories()
```
It might be cleaner to change the signature of `deleteExpiredRepositories()` to also return an `error`, so that you can just call `log15` once here and check the error instead of needing to have multiple calls to `log15` inside of `deleteExpiredRepositories()`. 

Example:


```go
func repositoryDeleter() {
	for {
		err := deleteExpiredRepositories()
        if err != nil {
           log15.Error("Error when evicting repositories", "error", err)
        }

	    time.Sleep(30 * time.Second)
	}
}
```

ggilmore commented on cmd/web-repo-proxy/main.go:
```diff
@@ -0,0 +1,323 @@
+	// filename strings are caught in the file creation stage.
+	for filename, _ := range request.FileContents {
+		if strings.Contains(filename, "/") {
+			return nil, errors.New("Repository filenames can't contain '/'")
```
I think we should use https://github.com/hashicorp/go-multierror so that all `filename` related errors are reported to the caller instead of just failing at the first one. You can see an example of how we use `multierror` here: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@3031fbab4617b94412edce2f6d2c75e1a674a5cb/-/blob/cmd/gitserver/server/cleanup.go#L133-165

So, this flow for this for loop could look something like:

```go
var errs *multierror.Error
for name, _ := request.FileContents {
   if strings.Contains(name, "/") {
     errs = multierror.Append(errs, fmt.Errorf("Filename %q is invalid, repository filenames can't contain '/'", name)
   }
}
return &request, errs.ErrorOrNil()
```

ggilmore commented on cmd/web-repo-proxy/main.go:
```diff
@@ -0,0 +1,323 @@
+		// Sanitize and error-check the request fields.
+		createRepositoryRequest, err := sanitizedCreateRequest(requestBody)
+		if err != nil {
+			log15.Warn("/create-repository request failed", "error", err)
```
`log15` allows you to reduce boilerplate by allowing you to create a new logger that includes a default key-value pair that you specify: https://godoc.org/github.com/inconshreveable/log15#New

So, at the top of `handleCreateRepository`, you could create a new logger, 

```go
logger := log15.New("route", "/create-repository")
```

And when you call methods on it later:

```go 
logger.Warn("Invalid request parameters", "error", err)
```

The log lines that it prints out will include the default key-value pair that you specified above, e.g:

```bash
WARN: Invalid request parameters, route: /create-repository, error: ...
```

ggilmore commented on cmd/web-repo-proxy/main.go:
```diff
@@ -0,0 +1,323 @@
+			return
+		}
+		// Empty repositories are considered invalid requests.
+		if len(createRepositoryRequest.FileContents) == 0 {
```
It seems strange that this check is performed separately from what's going on in `sanitizedCreateRequest` (checking to see whether or not the repo is empty doesn't seem to be all that different from checking the filenames.  Maybe this check should be included in `sanitizedCreateRequest()` as well.

ggilmore commented on cmd/web-repo-proxy/main.go:
```diff
@@ -0,0 +1,323 @@
+
+// Given a Reader with a create-repository request, return the corresponding
+// CreateRepositoryRequest with guaranteed-safe values, or an error.
+func sanitizedCreateRequest(requestBody io.Reader) (*CreateRepositoryRequest, error) {
```
I'll defer others on this one (cc @nicksnyder @slimsag), but it seems strange that this function accepts a such a large type like `io.Reader` given that:

1) We immediately attempt to decode a `CreateRepositoryRequest` from it and fail if that doesn't work
2) All the validation that we're doing operates on the actual struct that was parsed from the `io.Reader`

Could we change this function to accept a `CreateRepositoryRequest` and ask the caller to handle trying to decode it from the `io.Reader` beforehand?
